s3proxy/src/main.rs

```diff
@@ -0,0 +1,678 @@
+use anyhow::{Context, Result, bail};
+use axum::body::{Body, BodyDataStream};
+use axum::extract::State as AxumState;
+use axum::response::{IntoResponse, Response};
+use axum::{extract::Path, http::StatusCode, routing::get};
+use camino::{Utf8Path, Utf8PathBuf};
+use clap::Parser;
+use jsonwebtoken::{DecodingKey, Validation};
+use remote_storage::{DownloadError, DownloadOpts, GenericRemoteStorage, RemotePath};
+use serde::{Deserialize, Serialize};
+use std::sync::Arc;
+use std::time::Duration;
+use tokio_util::sync::CancellationToken;
+use tracing::{debug, error, info};
+
+const HELP: &str = "s3 proxy:
+If \"type\" is \"aws\", cli may look up the following environment variables:
+ AWS_ENDPOINT_URL
+ AWS_ACCESS_KEY_ID (note the lowercase, https://github.com/awslabs/aws-sdk-rust/issues/1020)
+ AWS_SECRET_ACCESS_KEY
+ or others, see https://docs.aws.amazon.com/sdkref/latest/guide/standardized-credentials.html
+In case of \"azure\", it may look up the following:
+ AZURE_STORAGE_ACCOUNT
+ AZURE_STORAGE_ACCESS_KEY";
+
+#[derive(clap::ValueEnum, Clone)]
+#[clap(rename_all = "kebab_case")]
+enum StorageType {
+    AWS,
+    Azure,
+}
+
+#[derive(Parser)]
+#[command(version, about, long_about = HELP)]
+struct Args {
+    #[arg(short, long)]
+    listen: std::net::SocketAddr,
+    #[arg(short, long)]
+    bucket: String,
+    #[arg(short, long)]
+    region: String,
+    #[arg(short = 't', long = "type")]
+    storage_type: StorageType,
+    #[arg(short, long, help = "Public key for verifying JWT tokens")]
+    pemfile: Utf8PathBuf,
+}
+
+#[tokio::main]
+async fn main() -> Result<()> {
```

Comment:
Maybe it's a matter of a personal choice and/or NIT, but can we think of a code structure right away, e.g.
- main file only contains the main func
- auth is a separate file
- router/http/handlers stuff is a separate file

To me personally, it'd be easier to review different parts. I'm not a fan of too many small files, but some reasonable breakdown usually makes sense, imo. WDYT?

Reply:
Fixed